db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,351 @@
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"sync"
+
+	sgbucket "github.com/couchbase/sg-bucket"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/google/uuid"
+)
+
+type AttachmentMigrationManager struct {
+	DocsProcessed base.AtomicInt
+	DocsChanged   base.AtomicInt
+	MigrationID   string
+	CollectionIDs []uint32
+	lock          sync.RWMutex
+}
+
+var _ BackgroundManagerProcessI = &AttachmentMigrationManager{}
+
+func NewAttachmentMigrationManager(metadataStore base.DataStore, metaKeys *base.MetadataKeys) *BackgroundManager {
+	return &BackgroundManager{
+		name:    "attachment_migration",
+		Process: &AttachmentMigrationManager{},
+		clusterAwareOptions: &ClusterAwareBackgroundManagerOptions{
+			metadataStore: metadataStore,
+			metaKeys:      metaKeys,
+			processSuffix: "attachment_migration",
+		},
+		terminator: base.NewSafeTerminator(),
+	}
+}
+
+func (a *AttachmentMigrationManager) Init(ctx context.Context, options map[string]interface{}, clusterStatus []byte) error {
+	newRunInit := func() error {
+		uniqueUUID, err := uuid.NewRandom()
+		if err != nil {
+			return err
+		}
+
+		a.MigrationID = uniqueUUID.String()
+		base.InfofCtx(ctx, base.KeyAll, "Attachment Migration: Starting new migration run with migration ID: %q", a.MigrationID)
+		return nil
+	}
+
+	if clusterStatus != nil {
+		var statusDoc AttachmentMigrationManagerStatusDoc
+		err := base.JSONUnmarshal(clusterStatus, &statusDoc)
+
+		// If the previous run completed, or there was an error during unmarshalling the status we will start the
+		// process from scratch with a new migration ID. Otherwise, we should resume with the migration ID, stats specified in the doc.
+		if statusDoc.State == BackgroundProcessStateCompleted || err != nil {
+			return newRunInit()
+		}
+		a.MigrationID = statusDoc.MigrationID
+		a.SetStatus(statusDoc.DocsChanged, statusDoc.DocsProcessed)
+		a.SetCollectionIDs(statusDoc.CollectionIDs)
+
+		base.InfofCtx(ctx, base.KeyAll, "Attachment Migration: Attempting to resume migration with migration ID: %s", a.MigrationID)
+
+		return nil
+	}
+
+	return newRunInit()
+}
+
+func (a *AttachmentMigrationManager) Run(ctx context.Context, options map[string]interface{}, persistClusterStatusCallback updateStatusCallbackFunc, terminator *base.SafeTerminator) error {
+	db := options["database"].(*Database)
+	migrationLoggingID := "Migration: " + a.MigrationID
+
+	persistClusterStatus := func() {
+		err := persistClusterStatusCallback(ctx)
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to persist cluster status on-demand for attachment migration operation: %v", migrationLoggingID, err)
```

Comment:
What does on-demand mean here? I don't know if cluster status makes sense here for logging since the cluster status doc is bark of BackgroundManager interface, I think this is really "Can't persist latest status for attachment migration"